firebase-database/src/androidTest/java/com/google/firebase/database/QueryTest.java

```diff
@@ -4578,22 +4578,188 @@ public void testGetWaitsForConnection() throws DatabaseException, InterruptedExc
+    FirebaseDatabase db = FirebaseDatabase.getInstance(app);
+    Semaphore semaphore = new Semaphore(0);
+    long val = 34;
+    DatabaseReference node =
+        db.getReference().child(Objects.requireNonNull(db.getReference().push().getKey()));
+    node.setValue(val);
+    ValueEventListener listener =
+        new ValueEventListener() {
+          @Override
+          public void onDataChange(@NonNull DataSnapshot snapshot) {
+            assertEquals(snapshot.getValue(), 34L);
+            semaphore.release();
+          }
+
+          @Override
+          public void onCancelled(@NonNull DatabaseError error) {
+            // no-op
+          }
+        };
+    node.addValueEventListener(listener);
+
+    try {
+      IntegrationTestHelpers.waitFor(semaphore);
+      DataSnapshot snapshot = await(node.get());
+      assertEquals(
+          snapshot.getValue(),
+          val); // TODO(mtewani): We might be getting this data straight from serverCache, so we may
+      // need to rewrite it
+    } catch (ExecutionException e) {
+      fail("get threw an exception: " + e);
+    }
+  }
+
+  public void testGetResolvesToCacheWhenOnlineAndChildLevelListener()
+      throws DatabaseException, InterruptedException {
+    FirebaseApp app =
+        appForDatabaseUrl(IntegrationTestValues.getDatabaseUrl(), UUID.randomUUID().toString());
+    FirebaseDatabase db = FirebaseDatabase.getInstance(app);
+    Semaphore semaphore = new Semaphore(0);
+    long val = 34;
+    DatabaseReference node =
+        db.getReference().child(Objects.requireNonNull(db.getReference().push().getKey()));
+    DatabaseReference childNode =
+        db.getReference().child(Objects.requireNonNull(node.push().getKey()));
+    node.setValue(val);
+    ValueEventListener listener = new ValueEventListener() { // TODO: move to child reference
+          @Override
+          public void onDataChange(@NonNull DataSnapshot snapshot) {
+            assertEquals(snapshot.getValue(), 34L);
+            semaphore.release();
+          }
+
+          @Override
+          public void onCancelled(@NonNull DatabaseError error) {
+            // no-op
+          }
+        };
+    childNode.addValueEventListener(listener);
+
+    try {
+      IntegrationTestHelpers.waitFor(semaphore);
+      DataSnapshot snapshot = await(node.get());
+      assertEquals(
+          snapshot.getValue(),
+          val); // TODO(mtewani): We might be getting this data straight from serverCache, so we may
+      // need to rewrite it
+    } catch (ExecutionException e) {
+      fail("get threw an exception: " + e);
+    }
+  }
+  /*
+   * Test:
+   *   1. Resolve to cache when offline
+   *   2. Resolve to non-cache when online
+   *   3. What happens when you have a listener on the same path as the get
+   *   4. What happens when you have a listener on the child path as the get
+   *   5. What happens when you have a listener on the parent path as the get
```

Comment:
We should also test these interactions with various filters (`startAt`, `limit` etc.) on the listen + get.

In particular, to verify that applyTaggedQueryOverwrite is working, we want to have a test with a disjoint listener and get at the same path and make sure they don't return overlapping data.